### PR TITLE
[LLD] [COFF] Pick timestamps from the SOURCE_DATE_EPOCH variable

### PR DESCRIPTION
**Pull request by mstorsjo**
The SOURCE_DATE_EPOCH environment variable can be set in order to get reproducible build.

When linking PE/COFF modules with LLD, the timestamp field is set to the current time, unless either the /timestamp: or /Brepro option is set. If neither of them is set, check the SOURCE_DATE_EPOCH variable, before resorting to using the actual current date and time.

### CODE REVIEW

File: lld/test/COFF/timestamp.test [hunk 1]
```diff
@@ -3,9 +3,11 @@ RUN: yaml2obj %p/Inputs/generic.yaml -o %t.obj
 RUN: lld-link %t.obj /debug /Brepro /entry:main /nodefaultlib /out:%t.1.exe
 RUN: lld-link %t.obj /debug /Brepro /entry:main /nodefaultlib /out:%t.2.exe
 RUN: lld-link %t.obj /debug /timestamp:0 /entry:main /nodefaultlib /out:%t.3.exe
+RUN: env SOURCE_DATE_EPOCH=0 lld-link %t.obj /debug /entry:main /nodefaultlib /out:%t.4.exe
```
Comment (MaskRay) [inline lld/test/COFF/timestamp.test:6]: It'd be useful to test another value like `clang/test/Preprocessor/SOURCE_DATE_EPOCH.c`

**Review by MaskRay — APPROVED** [review]

File: lld/test/COFF/timestamp.test (hunk 1 above)
Comment (mstorsjo) [reply lld/test/COFF/timestamp.test:6]: Thanks, I added more tests for more interesting cases.

What do you think of backporting this to 18.x? It’s not a bug fix per se, but a pretty safe improvement.

**Comment by aganea** [comment]
There’s not much info online about this. Do you know where it was first introduced?

**Comment by aganea** [comment]
Ah nvm found it: https://reproducible-builds.org/docs/source-date-epoch/

File: lld/test/COFF/timestamp.test (hunk 1 above)
Comment (MaskRay) [reply lld/test/COFF/timestamp.test:6]: LGTM for a backport to improve reproducibility! 

https://reproducible-builds.org/

**Review by MaskRay — APPROVED** [review]

**Comment by aganea** [comment]
Does this deserve an update of the release notes for 18.0?

**Comment by mstorsjo** [comment]
> Does this deserve an update of the release notes for 18.0?

Good point - yeah I think it does; I can make a separate PR to update the notes on the branch.

**Comment by MaskRay** [comment]
> > Does this deserve an update of the release notes for 18.0?
> 
> Good point - yeah I think it does; I can make a separate PR to update the notes on the branch.

The release notes for non-ELF ports are a barebone. @mstorsjo Do you want to find all interesting commits using `git log origin/release/17.x..origin/release/18.x -- lld/COFF`?

**Comment by mstorsjo** [comment]
> The release notes for non-ELF ports are a barebone. @mstorsjo Do you want to find all interesting commits using `git log origin/release/17.x..origin/release/18.x -- lld/COFF`?

I did something similar already - filling in the stuff which is already there - but I only looked at the commits that I have been involved in myself, that I'm more familiar with. But I guess it would be useful to have a look through all of them - it's just a little bit more work. I'll see if I can get to it in the next few days.

**Comment by DavidSpickett** [comment]
FYI: https://lab.llvm.org/buildbot/#/builders/178/builds/6797/steps/7/logs/FAIL__lld__timestamp_test

Only happens on the 32 bit bot.

**Comment by mstorsjo** [comment]
> FYI: https://lab.llvm.org/buildbot/#/builders/178/builds/6797/steps/7/logs/FAIL__lld__timestamp_test
> 
> Only happens on the 32 bit bot.

Oops, sorry about that, and sorry for not noticing the buildbot mail earlier.

It looks like the code itself is right, but `llvm-readobj` might not be able to print the value correctly on hosts with a 32 bit `time_t`.

Although, https://learn.microsoft.com/en-us/windows/win32/debug/pe-format doesn't seem to say anything explicitly about the signedness of this field either, but treating it as an unsigned seems reasonable. It does say that the value `0xFFFFFFFF` that we're using in the test doesn't represent a real timestamp. So perhaps we should test with a different value.

I guess the simplest way forward is to test with `0x7FFFFFFF` which should work safely everywhere. (And potentially test with `0xFFFFFFFE` if we know the host isn't limited to a 32 bit `time_t`?)

**Pull request merged**